### ISSUE TITLE
Incorrect analysis of package with shadowing import

### ISSUE THREAD

**Issue by judfs** [microsoft/pyright#6674]
**Describe the bug**
In the provided example, pyright will complain ` error: Module is not callable (reportGeneralTypeIssues)` about a valid line and will not complain about an invalid line a user might try as a "correction".

**Code or Screenshots**

```python
# run2.py
import my_package.foo
my_package.foo()
my_package.foo.foo()
```
```python
# my_package/__init__.py
from .foo import foo
```
```python
# my_package/foo.py
def foo():
    print("foo")
```

```
$ pyright run2.py 
/home/.../bug/run2.py
  /home/.../bug/run2.py:3:1 - error: Module is not callable (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations 
```
```
$ python3 run2.py 
foo
Traceback (most recent call last):
  File "run2.py", line 4, in <module>
    my_package.foo.foo()
AttributeError: 'function' object has no attribute 'foo'
```

Pyright complains about the valid `my_package.foo()` line and ignores the invalid `my_package.foo.foo()` line. This is bad because the 2nd option is the reasonable thing to write after seeing the first error in an editor.

**VS Code extension or command-line**
Same results cli and in vscode.
```
$ pyright --version
pyright 1.1.339
```

---
I acknowledge that there are other ways to write this code and have the same effect, either by changing `__init__` or _run2`. However a library I use, heavily uses the pattern in run2. This results in classes defined by that library being untyped as I edit my code because vscode/pylance/pyright gave up analyzing the library. If I were to edit the library with the invalid notation, then analysis for my code starts to "work".

#3131 is similar. It was closed as `as-designed`. However currently pyright correctly detects no issues in the example in that issue. Also I'm more concerned about the no error case than the wrong `Module is not callable`.

[Please edit the title for whatever the correct terminology is]

**Comment by erictraut** [comment]
Pyright is working as intended in this case, so I don't consider this a bug.

Your code is relying on fragile side effects in the Python import system that are difficult (and in many cases in possible) to model statically. If your intent is to access the local symbol `foo` from module `my_package`, you should use `import my_package` not `import my_package.foo`.
